internal/webhook/inferenceservice/mutating.go

```diff
@@ -0,0 +1,220 @@
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	webhookutils "github.com/opendatahub-io/opendatahub-operator/v2/pkg/webhook"
+)
+
+// create new type for connection types.
+type ConnectionType string
+
+const (
+	// ConnectionTypeURI represents uri connections.
+	ConnectionTypeURI ConnectionType = "uri-v1"
+	// ConnectionTypeS3 represents s3 connections.
+	ConnectionTypeS3 ConnectionType = "s3"
+	// ConnectionTypeOCI represents oci connections.
+	ConnectionTypeOCI ConnectionType = "oci-v1"
+)
+
+func (ct ConnectionType) String() string {
+	return string(ct)
+}
+
+type InferenceServingPath struct {
+	ModelPath           []string
+	ImagePullSecretPath []string
+}
+
+var IsvcConfigs = InferenceServingPath{
+	ModelPath:           []string{"spec", "predictor", "model"},            // used by S3, has map
+	ImagePullSecretPath: []string{"spec", "predictor", "imagePullSecrets"}, // used by OCI, has slice
+}
+
+//+kubebuilder:webhook:path=/platform-connection-isvc,mutating=true,failurePolicy=fail,groups=serving.kserve.io,resources=inferenceservices,verbs=create;update,versions=v1beta1,name=connection-isvc.opendatahub.io,sideEffects=None,admissionReviewVersions=v1
+//nolint:lll
+
+type ConnectionWebhook struct {
+	Client  client.Reader
+	Decoder admission.Decoder
+	Name    string
+}
+
+var _ admission.Handler = &ConnectionWebhook{}
+
+func (w *ConnectionWebhook) SetupWithManager(mgr ctrl.Manager) error {
+	hookServer := mgr.GetWebhookServer()
+	hookServer.Register("/platform-connection-isvc", &webhook.Admission{
+		Handler:        w,
+		LogConstructor: webhookutils.NewWebhookLogConstructor(w.Name),
+	})
+	return nil
+}
+
+func (w *ConnectionWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := logf.FromContext(ctx)
+
+	if w.Decoder == nil {
+		log.Error(nil, "Decoder is nil - webhook not properly initialized")
+		return admission.Errored(http.StatusInternalServerError, errors.New("webhook decoder not initialized"))
+	}
+
+	switch req.Operation {
+	case admissionv1.Create, admissionv1.Update:
+		// allowed connection types for connection validation on isvc.
+		allowedTypes := []string{
+			ConnectionTypeURI.String(),
+			ConnectionTypeS3.String(),
+			ConnectionTypeOCI.String(),
+		}
+
+		// validate the connection annotation
+		// - if has matched annoataion
+		// - if annaotation has valid value as that secret exists in the same namespace(permission allowed)
+		validationResp, shouldInject, secret, connectionType, decodedObj := webhookutils.ValidateConnectionAnnotation(ctx, w.Client, w.Decoder, req, allowedTypes)
```

Comment:
I would probably do the decoding outside of the `ValidateConnectionAnnotation`  method, so it is also easily testable

Reply:
let me update in the next round :D